Wikipedia/Code/NotificationsCenterViewController.swift

```diff
@@ -831,3 +835,32 @@ extension NotificationsCenterViewController {
         presentFiltersViewController()
     }
 }
+
+
+// MARK: - Device push permissions and silent Echo subscription
+
+extension NotificationsCenterViewController {
+
+    /// Mimics the Settings Push opt in approach, but Echo subscription fails silently to prevent interrupting the experience
+    func requestPushPermissionsAndSilentlySubscribeToEchoNotifications() {
+        deviceTokenRetryTask = RetryBlockTask { [weak self] in
```

Comment:
After testing this a bit, do you think we should exit early in `presentingOnboardingPushOptInIfNecessary` (don't show any prompt) if the user has permissions denied or `UserDefaults.standard.wmf_isSubscribedToEchoNotifications = true`? Both situations can happen if the user already interacted with app push settings before going to notifications center for the first time, and as a result nothing ends up happening when the user taps "Turn on push notifications", which could be confusing.

I know we discussed failing silently which is still good here in case of network failures or something, but if we can predict that it will fail here in these two cases, maybe it's better to not show the prompt at all.

Feel free to push back if this feels like too much of an edge case, or if it will over-complicate things too much. It was just something that jumped out at me.

Reply:
> do you think we should exit early in `presentingOnboardingPushOptInIfNecessary` (don't show any prompt) if the user has permissions denied or `UserDefaults.standard.wmf_isSubscribedToEchoNotifications = true`?

Good idea - updated!

Also, I missed that I forgot to subscribe to echo in the this flow if the device token is actually available 😅. Added that commit too!